### PR TITLE
Fix/format

### LINKED ISSUE #141: Consider relaxing JwtAlgorithm.DigSig
`JwtAlgorithm.DigSig` is defined as an enumeration. This limits the possible values that can be used and requires code changes to introduce support for new algorightms.

Given for instance no constants exist for `ESB256`, `ESB320`, `ESB384`, and `ESB512`, the following format fails to parse:
```json
{
  "format": {
    "mso_mdoc": {
      "alg": [
        "ES256",
        "ES384",
        "ES512",
        "EdDSA",
        "ESB256",
        "ESB320",
        "ESB384",
        "ESB512"
      ]
    }
  }
}
```

We must consider converting JwtAlgorithm.DigSig (or JwtAlgorithm) to a `value class`.

### LINKED ISSUE #142: Too strict  definition of Claim format designations
Presentation Exchange [contains](https://identity.foundation/presentation-exchange/spec/v2.0.0/#claim-format-designations) a section about the contents of an attribute named `format` which can be found either as a top-level element of presentation definition or with an `input_descriptor`.

This `format` is a JSON object having as attribute names claim formats coming by a [controlled registry](https://github.com/decentralized-identity/claim-format-registry/). In turn, each of these nested claim formats may point to another JSON Object which contains either an array `alg` or `proof_type` as defined in the [relevant](https://identity.foundation/claim-format-registry/schemas/presentation-definition-claim-format-designations.json) json schema.

Library reflects in its types the above rules, strictly. There are problems with this strictness:

OpenId4VP draft 20 uses claim formats that are not (yet) registered to the aforementioned Presentation Exchange controlled registry. In particular:

- `jwt_vc_json` & `jwt_vp_json` in [A.1.1](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#name-vc-signed-as-a-jwt-not-usin)
- `ac_vc` & `ac_vp` in [A.2](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#name-anoncreds)
- `mso_mdoc` in [A.3](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#name-iso-mobile-driving-license-)
- `sd+jwt-vc` to be added for `sd-jwt-vc` ([PR](https://github.com/openid/OpenID4VP/pull/115)). This doesn't only add a new value but redefines the contents of the relevant JSON to contain two arrays : `sd-jwt_alg_values` & `kb-jwt_alg_values`

So library-wise there are two options:
- Implement just the registered claims formats (current approach)
- Relax the `format` object accepting the fact that neither all keys are registered nor the contents of their value  

 The first is too restrictive, I guess. Perhaps we should go with the 2nd option.

### PR DESCRIPTION
**Pull request by dzarras**
~~Closes #141~~
Closes #142

### CODE REVIEW

File: src/main/kotlin/eu/europa/ec/eudi/prex/PresentationMatcher.kt [hunk 1]
```diff
@@ -89,7 +89,7 @@ typealias ClaimId = String
 
 interface Claim {
     val uniqueId: ClaimId
-    val format: ClaimFormat
+    val format: String
```
Comment (vafeini) [inline src/main/kotlin/eu/europa/ec/eudi/prex/PresentationMatcher.kt:92]: Format value class should be used here as format attribute is not any string but a string that can be converted to JsonObject

File: src/main/kotlin/eu/europa/ec/eudi/prex/Types.kt [hunk 2]
```diff
@@ -400,7 +265,7 @@ data class PresentationDefinition(
 @Serializable
 data class DescriptorMap(
     @Required val id: InputDescriptorId,
-    @Required val format: ClaimFormat,
+    @Required val format: String,
```
Comment (vafeini) [inline src/main/kotlin/eu/europa/ec/eudi/prex/Types.kt:268]: Format value class should be used here (format attribute is not any string but a string that can be converted to JsonObject) along with FormatSerializer

File: src/main/kotlin/eu/europa/ec/eudi/prex/PresentationMatcher.kt (hunk 1 above)
Comment (babisRoutis) [reply src/main/kotlin/eu/europa/ec/eudi/prex/PresentationMatcher.kt:92]: `ClaimFormat` used to be an enumeration. So I think in this case it's ok to replace it by a String

File: src/main/kotlin/eu/europa/ec/eudi/prex/Types.kt (hunk 2 above)
Comment (babisRoutis) [reply src/main/kotlin/eu/europa/ec/eudi/prex/Types.kt:268]: I don't think so [details](https://identity.foundation/presentation-exchange/spec/v2.0.0/#presentation-submission)

> The descriptor_map object MUST include a format property. The value of this property MUST be a string that matches one of the [Claim Format Designation](https://identity.foundation/presentation-exchange/spec/v2.0.0/#claim-format-designations). This denotes the data format of the [Claim](https://identity.foundation/presentation-exchange/spec/v2.0.0/#term:claim).

**Comment by babisRoutis** [comment]
Hi @dzarras 

@vafeini made an interesting comment. We can just relax the `Format` actually redefine it as per this PR. Should we delete though the previous types ?

That is should we permanently delete the types that we had or keep them for a future use?

**Review by babisRoutis — CHANGES_REQUESTED** [review]
@dzarras Since this is a major change in the public facade of the lib I sugges to go with a new "minor" version.

**Comment by dzarras** [comment]
@babisRoutis 
> @dzarras Since this is a major change in the public facade of the lib I sugges to go with a new "minor" version.

Should I bump to version to `0.3.0-SNAPSHOT`?

**Comment by dzarras** [comment]
> Should we delete though the previous types ? That is should we permanently delete the types that we had or keep them for a future use?

I would keep them only if we ended up defining `Format` as a sealed hierarchy of well-known (i.e. registered) formats, and generic (i.e. unregistered) formats.
This would mean extra maintenance burden, since we would have to keep the well-known formats aligned with the registry, and push minor updates whenever needed.

**Comment by babisRoutis** [comment]
> > Should we delete though the previous types ? That is should we permanently delete the types that we had or keep them for a future use?
> 
> I would keep them only if we ended up defining `Format` as a sealed hierarchy of well-known (i.e. registered) formats, and generic (i.e. unregistered) formats. This would mean extra maintenance burden, since we would have to keep the well-known formats aligned with the registry, and push minor updates whenever needed.

Indeed Dimitri. Let's keep the PR as is, just change the version to merge ti

**Pull request merged**